src/engines/misc/sends/sends.cpp

```diff
@@ -25,22 +29,79 @@ namespace otto::engines::sends {
     OTTO_UNREACHABLE;
   }
 
-  Sends::Sends() : audio(std::make_unique<Audio>()), screen_(std::make_unique<Screen>()), props{{*audio, *screen_}} {}
+  Sends::Sends(core::ui::Icon i) : audio(std::make_unique<Audio>()), screen_(std::make_unique<Screen>(i)), props{{*audio, *screen_}}
+  {
+    props.dry.on_change().connect([&](float a) {
+      if (is_recursive) return;
+      set(a, props.stored_levels.dry, props.stored_levels.fx1, props.stored_levels.fx2);
+    });
+    props.fx1.on_change().connect([&](float a) {
+      if (is_recursive) return;
+      set(a, props.stored_levels.fx1, props.stored_levels.dry, props.stored_levels.fx2);
+    });
+    props.fx2.on_change().connect([&](float a) {
+      if (is_recursive) return;
+      set(a, props.stored_levels.fx2, props.stored_levels.dry, props.stored_levels.fx1);
+    });
+  }
+
+  void Sends::step_scale(int steps)
+  {
+    set_scale(std::clamp(props.stored_levels.s + 0.01f * steps, 0.f, 1.f));
+  }
+
+  void Sends::set_scale(float x)
+  {
+    props.stored_levels.s = x;
+    is_recursive = true;
+    props.fx1 = props.stored_levels.fx1 * props.stored_levels.s;
+    props.fx2 = props.stored_levels.fx2 * props.stored_levels.s;
+    props.dry = props.stored_levels.dry * props.stored_levels.s;
+    is_recursive = false;
+  };
+
+  void Sends::set(float value, float& changed, float& other1, float& other2)
+  {
+    if (value < threshold && ((other1 < threshold && other2 < threshold) ||
+                              props.stored_levels.s < threshold)) { // Manually take care of "all 0" edge case
+      other1 = 1.f;
+      other2 = 1.f;
+      changed = 1.f;
+      props.stored_levels.s = 0.f;
+    } // Main function body below
+    else if (value >= other1 * props.stored_levels.s && value >= other2 * props.stored_levels.s) { // Now the largest
+      changed = 1;
+      other2 = other2 * props.stored_levels.s / value;
+      other1 = other1 * props.stored_levels.s / value;
+      props.stored_levels.s = value;
+    } else if (changed > other1 && changed > other2) { // Was largest, now not
+      float new_largest = std::max(other1, other2);
+      changed = value / (new_largest * props.stored_levels.s);
+      other1 = other1 / new_largest;
+      other2 = other2 / new_largest;
+      props.stored_levels.s = new_largest;
+    } else { // Was not the largest, remained so
+      changed = value / props.stored_levels.s;
+    }
+  }
```

Comment:
wait, scratch that, you can be really proud of it, this set function is extremely nice!

Reply:
thanks!